From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Wed, 30 Nov 2022 14:25:58 +0100
Subject: [PATCH] Skip entity move if movement is zero

License: MIT (https://opensource.org/licenses/MIT)
Gale - https://galemc.org

This patch is based on the following mixin:
"com/ishland/vmp/mixins/entity/move_zero_velocity/MixinEntity.java"
By: ishland <ishlandmc@yeah.net>
As part of: VMP (https://github.com/RelativityMC/VMP-fabric)
Licensed under: MIT (https://opensource.org/licenses/MIT)

// SYNC SOURCE START
rev: dc691c3a9af779db52b17c73047d94ff84dec020
branch: ver/26.2.0
repo: https://github.com/RelativityMC/VMP-fabric.git
track:
- "src/main/java/com/ishland/vmp/mixins/entity/move_zero_velocity/MixinEntity.java"
// SYNC SOURCE END

diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 5bac8fdbd18a4005e020aa8d63bcb469dc7d4835..1b85a87f02a9a0e0d54c366a5920595c7142b4a0 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -281,6 +281,7 @@ public abstract class Entity
     public float yRotO;
     public float xRotO;
     private AABB bb = INITIAL_AABB;
+    private boolean boundingBoxChanged = false; // Gale - VMP - skip entity move if movement is zero
     public boolean onGround;
     public boolean horizontalCollision;
     public boolean verticalCollision;
@@ -1138,7 +1139,7 @@ public abstract class Entity
     public void setOnGroundWithMovement(final boolean onGround, final boolean horizontalCollision, final Vec3 movement) {
         this.onGround = onGround;
         this.horizontalCollision = horizontalCollision;
-        this.checkSupportingBlock(onGround, movement);
+        this.checkSupportingBlock(onGround, movement); // Gale - VMP - skip entity move if movement is zero - diff on change
     }
 
     public boolean isSupportedBy(final BlockPos pos) {
@@ -1229,6 +1230,18 @@ public abstract class Entity
             }
             // Paper end
 
+            // Gale start - VMP - skip entity move if movement is zero
+            if (!this.boundingBoxChanged && delta.x == 0.0D && delta.y == 0.0D && delta.z == 0.0D) {
+                this.horizontalCollision = false;
+                this.verticalCollision = false;
+                this.verticalCollisionBelow = false;
+                this.minorHorizontalCollision = false;
+                this.setOnGroundWithMovement(false, false, Vec3.ZERO);
+                profiler.pop();
+                return;
+            }
+            this.boundingBoxChanged = false;
+            // Gale end - VMP - skip entity move if movement is zero
             delta = this.maybeBackOffFromEdge(delta, moverType);
             Vec3 movement = this.collide(delta);
             double movementLength = movement.lengthSqr();
@@ -1258,7 +1271,7 @@ public abstract class Entity
             if (movedVertically || this.isLocalInstanceAuthoritative()) {
                 this.verticalCollision = delta.y != movement.y;
                 this.verticalCollisionBelow = this.verticalCollision && delta.y < 0.0;
-                this.setOnGroundWithMovement(this.verticalCollisionBelow, this.horizontalCollision, movement);
+                this.setOnGroundWithMovement(this.verticalCollisionBelow, this.horizontalCollision, movement); // Gale - VMP - skip entity move if movement is zero - diff on change
             }
 
             if (this.horizontalCollision) {
@@ -4670,6 +4683,20 @@ public abstract class Entity
     }
 
     public final void setBoundingBox(final AABB bb) {
+        // Gale start - VMP - skip entity move if movement is zero
+        if (!this.boundingBoxChanged) {
+            AABB current = this.bb;
+            if (current != bb && (
+                    current.minX != bb.minX ||
+                    current.minY != bb.minY ||
+                    current.minZ != bb.minZ ||
+                    current.maxX != bb.maxX ||
+                    current.maxY != bb.maxY ||
+                    current.maxZ != bb.maxZ)) {
+                this.boundingBoxChanged = true;
+            }
+        }
+        // Gale end - VMP - skip entity move if movement is zero
         // CraftBukkit start - block invalid bounding boxes
         double minX = bb.minX,
                 minY = bb.minY,
